Drag drop

boombuler opened the pull request:
Hi,

I've started implementing system wide drag+drop for osx and windows. Please leave some comments!

Commit:
first bits of drag+drop support for osx

Commit:
raise routed events on drag and drop

boombuler committed:
implemented IDragData for OSX

kekekeks commented on src/Avalonia.Controls/DragDrop/DragDrop.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace Avalonia.Controls.DragDrop
+{
+    public sealed class DragDrop : AvaloniaObject
```
It could be a  static class. It also doesn't need to inherit from AvaloniaObject.

kekekeks commented on src/Avalonia.Controls/DragDrop/DragOperation.cs:
```diff
@@ -0,0 +1,13 @@
+namespace Avalonia.Controls.DragDrop
+{
+    [Flags]
+    public enum DragOperation
```
Might be worth to rename to [DragDropEffects](https://docs.microsoft.com/en-us/dotnet/api/system.windows.dragdropeffects?view=netframework-4.7.1)

boombuler replied:
Done.

kekekeks commented on src/Avalonia.Controls/DragDrop/IDragData.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Avalonia.Controls.DragDrop
+{
+    public interface IDragData
```
Might be worth to rename to [IDataObject](https://docs.microsoft.com/en-us/dotnet/api/system.windows.idataobject?view=netframework-4.7.1). We should probably also use it from our clipboard later.

boombuler replied:
Done.

kekekeks commented on src/Avalonia.Controls/DragDrop/DefaultDragDispatcher.cs:
```diff
@@ -0,0 +1,90 @@
+        private Interactive GetTarget(IInputElement root, Point local)
+        {
+            var target = root.InputHitTest(local)?.GetSelfAndVisualAncestors()?.OfType<Interactive>()?.FirstOrDefault();
+            if (target != null && DragDrop.GetAcceptDrag(target))
```
It should traverse the visual tree to check if there is an ancestor element, that accepts drag.

boombuler replied:
This is implemented like the behavior of WPF. Since the AcceptDrag property is inherited, all child elements should accept drags if not explicitly disabled, in that case it should not give the drag data to the parent elements. (IMHO)

kekekeks commented on src/OSX/Avalonia.MonoMac/TopLevelImpl.cs:
```diff
@@ -144,6 +151,75 @@ public void SetCursor(NSCursor cursor)
                 UpdateCursor();
             }
 
+            public override NSDragOperation DraggingEntered(NSDraggingInfo sender)
```
This logic shouldn't be in a platform-specific implementation. It should raise some events that are then handled by `TopLevel.cs.` Something like `public Action<RawDragEventArgs> OnDrag {get;set;}`

boombuler committed:
Initial Drag+Drop support for windows

boombuler committed:
DragDrop is now a static class

boombuler committed:
renamed DragOperation to DragDropEffects

boombuler committed:
corrected namespace. (copy+paste error)

boombuler committed:
renamed IDragData to IDataObject

boombuler committed:
reworked Drag event dispatching

boombuler committed:
reworked the dataobject for drag+drop

boombuler committed:
renamed DragDevice to DragDropDevice

boombuler committed:
fixed locator setup

boombuler committed:
Modified the ControlCatalog examples to run in STA ApartmentState

boombuler committed:
started dragging API

boombuler committed:
Basic generic + Windows Drag support.

boombuler committed:
set the cursor for the drop-effects

boombuler committed:
Added D+D StandardCursors

boombuler committed:
Improved DragSource

the drag source now gets the preferred drag effect and handles keyboard
inputs too

boombuler committed:
Allow Winforms + WPF compatible object drag

boombuler committed:
Prevent setting the cursor to often.

boombuler committed:
Merge branch 'master' into DragDrop

boombuler committed:
basic dragging support for OSX

boombuler committed:
OSX drag cleanup

boombuler committed:
drag file urls to finder.

boombuler committed:
automatically ensure dragged filenames are uris

grokys commented on samples/ControlCatalog.Desktop/Program.cs:
```diff
@@ -10,6 +10,7 @@ namespace ControlCatalog
 {
     internal class Program
     {
+        [STAThread]
```
Does this not work on .NET core? At the moment our templates assume one can use the same template across .NET Core and .NET Framework.

grokys replied:
This PR makes me think it should work on .NET core too: https://github.com/dotnet/coreclr/pull/15652

boombuler replied:
It does not work in the current version but as you mentioned the next release will take the attribute into account. For now only `Thread.CurrentThread.TrySetApartmentState(ApartmentState.STA);` will help on .net core

danwalmsley replied:
Couldn't this be set inside the appbuilder, so that doesn't have to be marked in user code?

boombuler replied:
[MSDN](https://msdn.microsoft.com/en-US/library/system.threading.thread.trysetapartmentstate(v=vs.110).aspx) says it needs to be done before the thread is started. might be possible but I guess Its to late.

What might be a possibility: Only register the Windows D+D implementation if the current thread is STA. That way "In-Process" D+D would always work and "Cross-Process" D+D would work on OSX even if the user did not set the ApartmentState.

kekekeks commented:
I am terribly sorry, hopefully will find time for a proper review this week.

grokys commented on src/Avalonia.Controls/DragDrop/Raw/RawDragEvent.cs:
```diff
@@ -0,0 +1,31 @@
+
+        public RawDragEvent(IDragDropDevice inputDevice, RawDragEventType type, 
+            IInputElement inputRoot, Point location, IDataObject data, DragDropEffects effects)
+            :base(inputDevice, GetTimeStamp())
```
This could simply be `: base(inputDevice, 0)` right?

grokys commented on src/Avalonia.Controls/DragDrop/DataObject.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace Avalonia.Controls.DragDrop
+{
+    public class DataObject : IDataObject
```
This class doesn't seem to be used anywhere. Am I missing something?

boombuler replied:
This class is an implementation which can be used to start dragging.
for example:

```csharp
DataObject dobj = new DataObject();
dobj.Set(DataFormats.Text, "Foobar");
var dragResult = await DragDrop.DoDragDrop(dobj, DragDropEffects.Copy|DragDropEffects.Move|DragDropEffects.Link);
```

grokys reviewed (changes requested):
Hi @boombuler - first of all _really_ sorry about the delay in reviewing this. Both @kekekeks and myself have found ourselves with little time recently.

I've taken a look at this and while I'm not the best person to comment on the platform-specific parts, I have a few comments:

- Drag/drop really needs to be implemented in `Avalonia.Input` rather than `Avalonia.Controls`. The only thing that seems to be preventing this is a dependency on `TopLevel`. To avoid this I'd suggest creating an `IInputRoot` or `IDragDropRoot` (not sure which would be best, @kekekeks @jkoritzinsky any feelings?) interface in `Avalonia.Input` and implementing it on `TopLevel`
- I tried getting it to work in ControlCatalog using the following code:

```
        public BorderPage()
        {
            this.InitializeComponent();
            DragDrop.SetAcceptDrag(this, true);
            AddHandler(DragDrop.DropEvent, Drop);
        }
```

However I couldn't get it to work. In WPF all I need to do is set `AllowDrop = true` and add a `Drop` handler. I assume the equivalent code here would be the above? What did I miss?
- Related to the above is `AcceptDrag` the same as WPF's `AllowDrop`? Why the different name?

Thanks for the work on this, it's really needed, and sorry again about the delay :(

boombuler committed:
Merge remote-tracking branch 'AvaloniaUI/master' into DragDrop

boombuler committed:
code cleanup

boombuler committed:
moved Drag+Drop sources to Avalonia.Input

Also the windows specific DragSource is
only used when the application runs as STA

boombuler commented:
Hi,

@grokys I moved most of the parts to `Avalonia.Input`. The only thing thats left in `Avalonia.Controls` is the "Fallback DragSource" which can be used on any platform which does not have a specific implementation but only works "In-Process" and not "Cross-Process"

The Windows specific DragSource is now only registered when the process runs as STA so it will at least work with in-process D+D when the process runs as MTA.

The sample code you have posted for the BorderPage example works fine on my machine. (Win32)
At least when I drag over one of the borders.

If you want I could rename `AcceptDrag` to `AllowDrop` I'm fine with both names.

grokys commented:
> The sample code you have posted for the BorderPage example works fine on my machine. (Win32)
At least when I drag over one of the borders.

Ah, you're right, I wasn't dragging over a border, d'oh.

grokys commented on src/Avalonia.Controls/Platform/DragSource.cs:
```diff
@@ -0,0 +1,212 @@
+
+namespace Avalonia.Platform
+{
+    class DragSource : IPlatformDragSource
```
So this is the "in-process" drag source that is used when there is no OS drag source available. It should probably be called `DefaultDragSource` or `InProcessDragSource` or something?

grokys commented on src/Avalonia.Input/DragDrop/DataObject.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace Avalonia.Input.DragDrop
+{
+    public class DataObject : IDataObject
```
I still don't see this class being used?

grokys reviewed (changes requested):
Looking good! I'm not able to test on OSX however, and my win32 knowledge is basic at best, so would be best if someone else could take a look too? If no-one else is available though, lets get this merged anyway.

Just a few small comments, and one request: do you think we could get some doc comments in the source for the public API? I can help out with that too if you like.

boombuler committed:
renamed DragSource to InProcessDragSource

boombuler committed:
Renamed AcceptDrag to AllowDrop and added docs

boombuler committed:
added Drag+Drop sample

boombuler committed:
don't use OSX-Platform DragSource

this should be used when the DragSource does work with any object.
For now only strings and filenames work.

boombuler commented:
Hi,

@grokys I've added a sample to the control-catalog which shows the usage of the `DataObject` which you mentioned. 
In addition to that: 
  - I've added some documentation
  - Disabled the OSX Platform DragSource until it is finished. (OSX can still recieve drag events from other applications)
  - Renamed AcceptDrag to AllowDrop to comply with WPF.

grokys commented:
@boombuler I think this looks ready to merge, however there's still WIP in the title. Is there still work to do on it?

boombuler commented:
Hi,

I've left the WIP because the OSX DragSource needs some work to send any object from one process to another. Its currently not in the service registration so OSX is only able to work with in-process d+d and receive infos from other processes. That can be done in a later PR (imho).
I would love to implement that but my trail version of rider expired...

boombuler committed:
Merge branch 'master' into DragDrop
